КП АП СРНС. Этап 3

ZherebinVR opened the pull request:
Начал делать 3 часть КП. Столкнулся со сложностью запуска проекта через CMake. Как я понял, после нескольких недель попыток, CMake заточена под Unix-подобные ОС и с вин10 она не дружит. Танцы с бубнами тоже не помогают. От сюда возникают 2 вопроса: 
1. Возможно ли на вин10 запустить проект через CMake, и если можно, то как? 
2. Можно ли делать проект в другой сборке, по другим правилам и тд.?

ZherebinVR committed:
Starting C++

ZherebinVR committed:
test new branch

ZherebinVR committed:
Создан рабочий проект

ZherebinVR committed:
Создан проект тестов

ZherebinVR committed:
Добавлены структуры.
Добавлен класс эфемерид в струкруты.
Добавлен расчет время ГЛОНАСС

ZherebinVR committed:
Добавлены функции GMST_calc, CrdTrnsf2Inertial.
student.h(!?) переименовано в structures.h

ZherebinVR commented:
Текущие наработки у меня в ветке qmake_branch

ZherebinVR committed:
Добавлено численное интегрирование дифуров методом Рунге-Кутты.
Осталось только не споткнуться об это бревно

Korogodin commented:
1. CMake с win10 работать может, другое дело - мои скрипты. Никто их никогда под Win не запускал и не ориентировал. Как сказано в описании к КП, проще всего - установить виртуалку с linux-системой
2.Да, можно настроить свою сборку

Korogodin commented:
Перед тем, как переносить РК на Си, я бы сначала реализовал его в матлабе (без ode).

ZherebinVR committed:
Реализация метода Рунгы-Кутты 4-го порядка в матлабе

ZherebinVR commented:
Реализовал. Скайвью с моей реализацией выглядит так же, как и с ode:
![image](https://user-images.githubusercontent.com/55599416/81650145-1afca380-943a-11ea-91e0-9c115c9a3452.png)

ZherebinVR committed:
merge from master

ZherebinVR committed:
Добавлен расчет количества отсчетов.
Переведение к порядку структурок.

Korogodin commented:
Теперь надо:
1. Перекидать почти дословно код в Си
2. Выгрузить из матлаба входной/выходной датасет
3. Отладить Си, чтоб на тестовом датасете давал тот же результат
4. Оптимизировать
5. Проверить на утечки

ZherebinVR committed:
Исправления Рунгы-Кутты

ZherebinVR committed:
Добавлено формирование выходного массива Yout

ZherebinVR committed:
Добавлена запись результатов в файл.
Добавлен учет ускорений.
В матлабе добавлена загрузка и отоброжение результатов из С++

ZherebinVR commented:
Возникли странности с отладкой. Все формулы верно считают, но с увеличение числа итераций очень сильна растет ошибка. Для наглядности загрузил результат в матлаб:
![image](https://user-images.githubusercontent.com/55599416/81691876-38476700-9466-11ea-9a04-693830f99318.png)

ZherebinVR committed:
Коррекция ошибок

Korogodin commented:
Вот тут тебе и поможет матлаб. При double'вом представлении чисел в Си разница с матлабом должна быть не больше 10^-14

Прям теперь берешь и по-шагам сравниваешь матлаб с Си, осознаешь источник проблем

ZherebinVR commented:
Источник проблем найден и устранен. Результат "совпадает" с моделью:
![image](https://user-images.githubusercontent.com/55599416/81814411-6d6cbb80-9531-11ea-8a42-82ca8dba71f0.png)

ZherebinVR committed:
Исправлена ошибка в численном интегрировании.

Korogodin commented:
Определенно стало лучше, но совпадает или нет - в таком масштабе не видно

ZherebinVR commented:
Взял разницу координат модели и С, вывел на график:
![image](https://user-images.githubusercontent.com/55599416/81962942-1f34e680-961d-11ea-8b9a-9fef921427eb.png)
Разница в +-2км считается нормальной?) 
По заданию требуется провести расчет с шагом 0.1с. Программа на С с этим спокойно справляется, а вот матлаб считает целую вечность - за 5 минут только 10% точек посчитал.

ZherebinVR committed:
Перенос основного расчета из main.cpp в glnsvpos.cpp.
Перенос всех вспомогательных функций в func.cpp.
В модели добавлен вывод разницы координат.

ZherebinVR committed:
del temp file

ZherebinVR committed:
Merge branch 'master' into qmake_branch

# Conflicts:
#	simulation/RungeKutta4.m

Korogodin commented:
1. О расхождении координат. Как я уже писал, смириться можно с 1e-14 от значений, в данном случае это 0.2 микрона. Да, эта ошибка будет интегрироваться и нарастать, но у тебя такой точности нет и в базовой точке, где интегрирования нет
2. О времени расчетов. Такого быть не должно, обычно разница только в разы. Вероятно, ты в матлабе где-то перерастягиваешь массивы или т.п.

Korogodin commented:
Влад, как продвижение? 

В коде в глаза ничего не бросается. Как я писал выше, самый простой способ - запустить параллельно матлаб и Си (дебаггер) и идти по шагам, сравнивая числа. Уже в начальной точке у тебя по какой-то причине расхождения.

ZherebinVR commented:
Пытался добавить тесты, но безуспешно. 
Расхождение наблюдаются из-за учета ускорений. Вот так выглядит ошибка с учетом ускорений:
![image](https://user-images.githubusercontent.com/55599416/82209163-7be21b00-9915-11ea-9b49-0f008c44ab67.png)
Вычислял так D_DeltaF1 = sqrt( DeltaF1(:,1).^2 + DeltaF1(:,2).^2 + DeltaF1(:,3).^2 );
А вот без учета ускорений:
![image](https://user-images.githubusercontent.com/55599416/82209586-29edc500-9916-11ea-9d34-9395a33d8f67.png)
Видно, что ошибка имеет случайный характер, и не превышает 9 м, но все-равно это на порядки больше чем 0,2 мкм
UPDATE: Исправил ошибку в расчетах ускорения, теперь ошибка выглядит как нижняя картинка.

Korogodin commented:
Ты пытаешься смотреть на проблему свысока, надо иногда опускать ниже. Ещё раз намекаю на очевидное решение: так как Си у тебя повторяет матлаб, то достаточно запустить первое и второе в режиме дебаггера и пройтись по цепочке, сравнивая значения. 

Сейчас ошибка у тебя не нарастает со временем. И в ней прослеживается дискретный характер. Такое может быть, например, при использовании разной точности представления чисел в уравнениях для трансформации координат. Или какой-нибудь банальный printf для построения графика, в котором у тебя выводится недостаточное количество цифр после запятой.

ZherebinVR commented:
Точно! Это все из-за printf. Выводилось недостаточно знаков после запятой. Спасибо за напоминание!!)
Поставил вывод 15 цифр после запятой, результат такой:
![image](https://user-images.githubusercontent.com/55599416/82218955-552be080-9925-11ea-8587-f0805c6c20ca.png)

Korogodin commented:
Во, другое дело! Теперь автоматизировать - чтобы компьютер получал этот график и выносил решение, и тест готов. Для этого из матлаба выгрузи выходные (а лучше и входные) данные, а затем используй их в программе. Только опять же, не забудь про точность представления чисел)

ZherebinVR committed:
Исправлена ошибка учета ускорений.
Добавлены новые функции записи структуры Y_s write_struct_Y и чтения read_struct_Y.
Коррекция обозначений и общего офрмления проекта.

ZherebinVR committed:
Начало Тестов положено

ZherebinVR committed:
Успешный запуск тестов

ZherebinVR committed:
Настроен запуск функций add() и mult() из основного проекта

ZherebinVR commented:
Настроил юнит тесты помощью среды тестирования Тест Boost:
![image](https://user-images.githubusercontent.com/55599416/82572280-c91bf200-9b8c-11ea-8970-8ec455bbee9f.png)
Результат выполнения следующий: 
![image](https://user-images.githubusercontent.com/55599416/82572311-d76a0e00-9b8c-11ea-87ff-30cf14e01400.png)
Тест функции add показал ошибку, так как в саму функцию внесена ошибка, а время выполнения 823 мкс. Тест функции mult завершился успешно за 22 мкс. Суммарное время тестирования составило 4187 мкс. 
Правильные ли я сделал выводы из полученных результатов? Адекватные ли результаты?

ZherebinVR committed:
Настроен тест для функции NT_calc()

ZherebinVR commented:
Добавил тест для проверок утечек памяти:
![image](https://user-images.githubusercontent.com/55599416/82602774-4742bd80-9bba-11ea-9a8e-268629208e63.png)
С помощь утилиты Dr.Memory проверил, обнаружил утечку на 4800 байт.
![image](https://user-images.githubusercontent.com/55599416/82602888-68a3a980-9bba-11ea-83f7-871c88b784ff.png)
Если все ок, то сетап для финального рывка готов

Korogodin commented:
Всё так, теперь эти скиллы надо применить к нашей библиотеке: написать тест для сравнения Си с матлабом, заодно на нем можно и утечки проверить

Korogodin commented:
Пора финалить, на следующей неделе у нас защита

ZherebinVR committed:
Тесты, тесты и еще раз тесты, и немного про утечки
Отчет по 3 части.

ZherebinVR commented:
Оформил отчет по 3 этапу, остался только общий вывод по курсовому

ZherebinVR committed:
Финалочка

ZherebinVR commented:
Оформление курсового завершено

Korogodin commented:
1. Вчера выяснилось, что это курсовая "работа", а не проект. Я всех обманывал) Поправь, пожалуйста, на титуле
2. Проще было написать "2.0"
![image](https://user-images.githubusercontent.com/747285/83325633-66a2b000-a276-11ea-8842-5d2f00462175.png)
3. Не стал тратить время и разбираться, вероятно, так и надо, но декабрь не используется
![image](https://user-images.githubusercontent.com/747285/83325676-b8e3d100-a276-11ea-86b3-fbf2bbb450f9.png)
4. "Список использованных источников"
![image](https://user-images.githubusercontent.com/747285/83326128-10d00700-a27a-11ea-9a07-6b4ff447c543.png)

В целом, отлично!

Korogodin commented:
3 июня с утра назначена защита

ZherebinVR commented:
Поправил титульный и "Список использованных источников", надо это пушить куда-то?
По защите будет более подробная информация? Где, во сколько, к чему готовиться и т.д.?

Korogodin commented:
Можешь ещё один ПР завести

По защите: я+Куликов в 10 онлайн. С меня спрашивали только время, не я организую. 
Вероятно, скайп. Будем вас по-одному приглашать и расспрашивать, я так себе это вижу
